clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -229,7 +229,7 @@ public long connectionDelay(Node node, long now) {
      */
     @Override
     public boolean connectionFailed(Node node) {
-        return connectionStates.connectionState(node.idString()).equals(ConnectionState.DISCONNECTED);
+        return connectionStates.isDisconnected(node.idString());
```

Comment:
Just want to understand the weirdness here. Is it because `isDisconnected` could mean either "used to be connected" or "never connected"? In our current context it means the former one. I am not sure if we want to keep a separate set of supposedly-connected, which introduces another state to maintain.

We can add a new state (something like `NO_STATE`) to ConnectionState to indicate "never connected" and let ConsumerNetworkClient to use it if that would help disambiguate.